browser/ui/webui/settings/pin_shortcut_handler.cc

```diff
@@ -38,20 +38,63 @@ void PinShortcutHandler::HandleCheckShortcutPinState(
     const base::Value::List& args) {
   AllowJavascript();
 
+  CheckShortcutPinState(false);
+}
+
+void PinShortcutHandler::CheckShortcutPinState(bool from_timer) {
   shell_integration::IsShortcutPinned(
       base::BindOnce(&PinShortcutHandler::OnCheckShortcutPinState,
-                     weak_factory_.GetWeakPtr()));
+                     weak_factory_.GetWeakPtr(), from_timer));
 }
 
+#if BUILDFLAG(IS_WIN)
+void PinShortcutHandler::OnPinStateCheckTimerFired() {
+  CheckShortcutPinState(true);
+}
+#endif
+
 void PinShortcutHandler::OnPinShortcut(bool pinned) {
   DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
 
+#if BUILDFLAG(IS_WIN)
+  // On Windows, ignore |pinned| value as Windows asks to user via another
+  // os notification to pin. |pinned| state just represents that api call is
+  // succuss. So, polling to check whether user allowed or not.
+  if (!pin_state_check_timer_) {
+    pin_state_check_timer_ = std::make_unique<base::RetainingOneShotTimer>(
+        FROM_HERE, base::Seconds(2),
+        base::BindRepeating(&PinShortcutHandler::OnPinStateCheckTimerFired,
+                            weak_factory_.GetWeakPtr()));
+  }
+  // We'll try to check 10 times after asking to pin to Windows.
+  // Maybe user will see os notification for pin right after clicking the Pin
+  // button. If user doesn't click allow in 20s, we'll not monitor anymore as we
+  // could give proper pin state after reloading.
+  pin_state_check_count_ = 10;
+  pin_state_check_timer_->Reset();
+#else
   NotifyShortcutPinStateChangeToPage(pinned);
+#endif
 }
 
-void PinShortcutHandler::OnCheckShortcutPinState(bool pinned) {
+void PinShortcutHandler::OnCheckShortcutPinState(bool from_timer, bool pinned) {
   DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
 
+#if BUILDFLAG(IS_WIN)
+  if (from_timer) {
+    pin_state_check_count_--;
+    // Clear if pinned or we did all try.
+    if (pinned || pin_state_check_count_ == 0) {
+      pin_state_check_count_ = 0;
+      pin_state_check_timer_.reset();
+    } else {
+      // Check again.
+      pin_state_check_timer_->Reset();
```

Comment:
How about using RepeatingTimer to remove this line?

Reply:
I think `RetainingOneShotTimer` is more fit here.
When it's fired, we need to post async call to get pinned state.
And then start timer again after getting async call's result.
If `RepeatingTimer` is used, next one is fired before checking pinned state.

Reply:
Aha, you're right. Thanks for explaning.